### ISSUE TITLE
set_epoch for DistributedSampler

### ISSUE THREAD

**Issue by ananyahjha93** [williamFalcon/pytorch-lightning#224]
**Describe the bug**
PyTorch example suggests the use set_epoch function for DistributedSampler class before each epoch start. I could not find this function call in lightning's trainer module. 

https://github.com/pytorch/examples/blob/master/imagenet/main.py
Line 232-234

As can be seen from the DistributedSampler class code (https://github.com/pytorch/pytorch/blob/master/torch/utils/data/distributed.py), the set_epoch function is required to set the seed for each __iter__ function call.

Can you confirm if this function has been called on DistributedSampler (for training dataset) at some point in lightning's trainer module?

**Comment by williamFalcon** [comment]
it's not called which means it is equivalent to shuffle=False (ie: shuffled once but not again every epoch). We can add it though to enable shuffling

**Comment by williamFalcon** [comment]
do you want to submit that PR?

**Comment by ananyahjha93** [comment]
Added a PR for this, do you want me to update any docs in my commit related to this issue? Also, I have successfully tested the library with this fix included in my experiments.

**Comment by airsplay** [comment]
> it's not called which means it is equivalent to shuffle=False (ie: shuffled once but not again every epoch). We can add it though to enable shuffling

A short note on this for future readers: 
No setting epochs in DistributedSampler does not equivalent to `shuffle=False` (for both dataloader and distributed sampler).

**Comment by xuanyang19** [comment]
How could I called set_epoch() on DistributedSampler in pytorch lightning? Although I set distributedsampler(shuffle = True), the train index is always the same at every epoch. What should I do to make the train index become different at every epoch?

**Comment by KeplerWang** [comment]
> How could I call set_epoch() on DistributedSampler in pytorch lightning? Although I set distributedsampler(shuffle = True), the train index is always the same at every epoch. What should I do to make the train index become different at every epoch?

@xuanyang19 Hi! I got the same question these days. Better late than never. According to my test, set_epoch() is now automatically executed for train_dataloader (even when shuffle=False) in pytorch-lightning 2.0.7.
